client/asset/eth/eth.go

```diff
@@ -1000,8 +966,29 @@ func (eth *ExchangeWallet) findSecret(ctx context.Context, secretHash [32]byte,
 
 // Refund refunds a contract. This can only be used after the time lock has
 // expired.
-func (*ExchangeWallet) Refund(coinID, contract dex.Bytes, feeSuggestion uint64) (dex.Bytes, error) {
-	return nil, asset.ErrNotImplemented
+func (eth *ExchangeWallet) Refund(_, contract dex.Bytes, feeSuggestion uint64) (dex.Bytes, error) {
+	version, secretHash, err := dexeth.DecodeContractData(contract)
+	if err != nil {
+		return nil, fmt.Errorf("Refund: failed to decode contract: %w", err)
+	}
+
+	refundable, err := eth.node.isRefundable(secretHash, version)
+	if err != nil {
+		return nil, fmt.Errorf("Refund: failed to check isRefundable: %w", err)
+	}
+	if !refundable {
+		return nil, fmt.Errorf("Refund: swap with secret hash %x is not refundable", secretHash)
+	}
+
+	tx, err := eth.node.refund(eth.ctx, secretHash, feeSuggestion, version)
+	if err != nil {
+		return nil, fmt.Errorf("Refund: failed to call refund: %w", err)
+	}
```

Comment:
This uses the server's feeSuggestion as the maxFee.

So... With the new dynamic txs, should the suggestion be the TipCap? We cannot choose the base fee, and I personally think the overall max should be the user's chosen max in all places.

Correct me if my understand is wrong. You have the base fee which is out of our control, must be paid. You have the tip or priority fee which we should put at 2 gwei basically, but this is the fee as we usually know it. Then we can choose the max that base plus tip cannot go over.

Reply:
For refunds, it would be fine to ignore server's feeSuggestion for `GasFeeCap` (max base fee) and instead use the user's specified own.  But this arg to `(*nodeClient).refund` ends up in `bind.TransactionOpts.GasFeeCap`, and the tip is requested on the fly by `txOpts` > `netFeeState`.

Regarding the units of the server's provided fee rates, the server's `(*Backend).FeeRate` returns a suggestion that is (*current* base + suggested fee tip cap).  See the call to `suggestGasPrice` that is handled by `(*PublicEthereumAPI).GasPrice` in internal/ethapi/api.go.    This is documented as "for legacy transactions" for obvious reasons -- it sums the two rates.  This does raise some questions.

First is that we've talked about the server-prescribed and suggested rates, which find their way to the `GasFeeCap` for dynamic client txns, being based on some **multiple of the current base fee rate** (e.g. 125%) rather than just adding the current suggested tip to the current base fee, like the `PublicEthereumAPI` method does as described above.  We only have to look at how base fee swings from one block to the next

Second is that for the purposes of the init txns, the prescribed fee rate for matches must currently refer to the max gas fee cap because if the user's unconfirmed init txn specifies any less, then it should rejected.  See how `server/asset/eth.(*Backend).newSwapCoin` uses `tx.GasPrice()`, which for dynamic txns returns `GasFeeCap`.  A related shower thought is that the server's `(*Backend).FeeRate` method could return a second fee value so the tip cap can also be enforced, but I have a feeling just the gas cap is fine.

Reply:
Sorry, that was almost completely off-topic.  This is the Refund PR, and yes the `(*nodeClient).refund` method can be changed to use `feeSuggestion` only as a fallback or not at all, if it determines a reasonable base fee cap locally.